src/andromede/model/resolve_library.py

```diff
@@ -60,55 +60,100 @@ def resolve_library(
         for preloaded_lib in preloaded_libs:
             preloaded_port_types.update(preloaded_lib.port_types)
 
-    output_lib = Library(port_types=preloaded_port_types, models={})
+    output_lib_dict: Dict[str, Library] = (
+        dict((l.id, l) for l in preloaded_libs) if preloaded_libs else {}
+    )
 
-    todo: List[str] = list(yaml_lib_dict)
-    done: Set[str] = set()
+    remaining_lib_ids: List[str] = list(yaml_lib_dict)
+    treated_lib_ids: Set[str] = set()
     import_stack: List[str] = []
 
-    while todo:
-        next_lib_id = todo.pop()
+    while remaining_lib_ids:
+        next_lib_id = remaining_lib_ids.pop()
 
-        if next_lib_id in done:
+        if next_lib_id in treated_lib_ids:
             continue
         else:
             import_stack.append(next_lib_id)
 
         while import_stack:
-            cur_lib = yaml_lib_dict[import_stack[-1]]
-            dependencies = set(cur_lib.dependencies) - done
+            cur_yaml_lib = yaml_lib_dict[import_stack[-1]]
+            current_lib = Library(id=cur_yaml_lib.id, port_types={}, models={})
+
+            # Add already parsed port types from dependencies in current lib
+            _add_preloaded_port_types_to_current_lib(preloaded_port_types, current_lib)
+            _add_treated_dependent_port_types_to_current_lib(
+                output_lib_dict, treated_lib_ids, cur_yaml_lib, current_lib
+            )
 
-            if dependencies:
-                first_dependency = dependencies.pop()
+            remaining_dependencies = set(cur_yaml_lib.dependencies) - treated_lib_ids
 
-                if first_dependency in import_stack:
-                    raise Exception("Circular import in yaml libraries")
-                import_stack.append(first_dependency)
+            if remaining_dependencies:
+                _add_dependencies_to_stack(import_stack, remaining_dependencies)
 
             else:
-                port_types = [_convert_port_type(p) for p in cur_lib.port_types]
-                port_types_dict = dict((p.id, p) for p in port_types)
+                _treat_lib(current_lib, cur_yaml_lib, output_lib_dict)
+                _update_treated_libs_and_import_stack(treated_lib_ids, import_stack)
+
+    return output_lib_dict
+
+
+def _add_preloaded_port_types_to_current_lib(
+    preloaded_port_types: dict[str, PortType], current_lib: Library
+) -> None:
+    current_lib.port_types.update(preloaded_port_types)
+
+
+def _add_treated_dependent_port_types_to_current_lib(
```

Comment:
maybe "handle" or "process" better than "treat" (or resolve to keep the wording).

Reply:
Done